mesonbuild/cargo/version.py

```diff
@@ -0,0 +1,90 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright © 2022-2023 Intel Corporation
+
+"""Convert Cargo versions into Meson compatible ones."""
+
+from __future__ import annotations
+import typing as T
+
+
+def convert(cargo_ver: str) -> T.List[str]:
+    """Convert a Cargo compatible version into a Meson compatible one.
+
+    :param cargo_ver: The version, as Cargo specifies
+    :return: A list of version constraints, as Meson understands them
+    """
+    # Cleanup, just for safety
+    cargo_ver = cargo_ver.strip()
+    cargo_vers = [c.strip() for c in cargo_ver.split(',')]
+
+    out: T.List[str] = []
+
+    for ver in cargo_vers:
+        # This covers >= and =< as well
+        # https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#comparison-requirements
+        if ver.startswith(('>', '<', '=')):
+            out.append(ver)
+
+        elif ver.startswith('~'):
+            # Rust has these tilde requirements, which means that it is >= to
+            # the version, but less than the nxt version
+            # https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#tilde-requirements
+            # we convert those into a pair of constraints
+            v = ver[1:].split('.')
+            out.append(f'>= {".".join(v)}')
+            if len(v) == 3:
+                out.append(f'< {v[0]}.{int(v[1]) + 1}.0')
+            elif len(v) == 2:
+                out.append(f'< {v[0]}.{int(v[1]) + 1}')
+            else:
+                out.append(f'< {int(v[0]) + 1}')
+
+        elif ver.startswith('^'):
+            # https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#caret-requirements
+            out.append(f'== {ver[1:]}')
```

Comment:
There seems to be a misunderstanding here. A requirement of `^1.2.3` does not mean `= 1.2.3`, it means `>=1.2.3, <2.0.0`.

Reply:
Yeah, I was just about to comment that I discovered the same thing when talking to someone else.

The current logic is therefore in error.

Reply:
Eh, looks pretty well documented to me. Almost everybody specifies requirements as `1.2.3` in the Rust world. I think I've seen a single instance of someone using `^1.2.3`.

Reply:
yeah, that documentation is terribly worded "^1.2.3 is exactly equivalent to 1.2.3" sure sounds like `^1.2.3.` means `exactly 1.2.3`

Reply:
> Almost everybody specifies requirements as 1.2.3 in the Rust world. I think I've seen a single instance of someone using ^1.2.3.

That's probably true for Cargo.toml files, but crates.io uses it everywhere: https://crates.io/crates/gstreamer/0.20.5/dependencies.

Anyway, the fix should be trivial, drop first `^` char and fallthrought

Reply:
ah, fixed already.

Reply:
> yeah, that documentation is terribly worded "^1.2.3 is exactly equivalent to 1.2.3" sure sounds like `^1.2.3.` means `exactly 1.2.3`

Ah, I see where you're coming from. If you didn't read the section above and assume that `1.2.3` means `= 1.2.3`, it can be confusing with the current wording.
What about this?:
> [...]. `^1.2.3` is exactly equivalent to `1.2.3`, which is documented above.

Would that resolve the ambiguity from your perspective?

> That's probably true for Cargo.toml files, but crates.io uses it everywhere: https://crates.io/crates/gstreamer/0.20.5/dependencies.

Fair. Forgot about that.